ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -390,8 +390,13 @@ static void sendDataStreamException(Throwable throwable, DataStreamRequestByteBu
   void cleanUp(Set<ClientInvocationId> ids) {
     for (ClientInvocationId clientInvocationId : ids) {
       Optional.ofNullable(streams.remove(clientInvocationId))
-          .map(StreamInfo::getLocal)
-          .ifPresent(LocalStream::cleanUp);
+          .ifPresent(streamInfo -> {
+            streamInfo.getDivision()
+                .getDataStreamMap()
+                .remove(clientInvocationId);
+            streamInfo.getLocal().cleanUp();
+            streamInfo.applyToRemotes(out -> out.out.closeAsync());
+          });
```

Comment:
Let's use the `removeDataStream` method here.

BTW, we should update it to not passing `StreamInfo` and then check if the removed steam is the same.  If not, call `cleanUp(..)`.  See below.
```java
-  private void removeDataStream(ClientInvocationId invocationId, StreamInfo info) {
+  private StreamInfo removeDataStream(ClientInvocationId invocationId) {
     final StreamInfo removed = streams.remove(invocationId);
-    if (info == null) {
-      info = removed;
-    }
-    if (info != null) {
-      info.getDivision().getDataStreamMap().remove(invocationId);
-      info.getLocal().cleanUp();
+    if (removed != null) {
+      removed.cleanUp(invocationId);
     }
+    return removed;
   }
 
   private void readImpl(DataStreamRequestByteBuf request, ChannelHandlerContext ctx,
@@ -479,7 +480,10 @@ public class DataStreamManagement {
       try {
         if (exception != null) {
           replyDataStreamException(server, exception, info.getRequest(), request, ctx);
-          removeDataStream(key, info);
+          final StreamInfo removed = removeDataStream(key);
+          if (removed != info) {
+            info.cleanUp(key);
+          }
         }
       } finally {
         request.release();
```

Reply:
@szetszwo Thanks for the review.

```
+          if (removed != info) {
+            info.cleanUp(key);
+          }
```
For the above code, I was wondering in what cases will "removed not equal to info". I think in the original code, info was checked because in catch exception in `read()` we don't know the info and we give an "null", but here `info` should be equal to r`emoved`?

Reply:
Updated, PTAL.